### ISSUE TITLE
add a field to build manifest file for minimum zig version

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#14475]
This proposal is to add a field to build.zig.zon like this:

```zig
    .minimum_zig_version = "0.11.1",
```

This would provide a potential way for high level tooling to satisfy zig version in whatever way deemed necessary. I would foresee, for example, an IDE extension using this field to find the correct version of zig to download and install in order to build a project.

At the very least it could help users by providing a compile error if the current zig version is too old, which may lead to a better outcome than if the user observed a confusing compile error.

**Comment by rofrol** [comment]
What about max version?
Maybe specifying like in elm.json:

`"elm-version": "0.19.0 <= v < 0.20.0",`

https://github.com/elm/compiler/blob/047d5026fe6547c842db65f7196fed3f0b4743ee/docs/elm.json/package.md

**Comment by marler8997** [comment]
> what about a max version?

Good question.  I talked about this a little here in the PR that this proposal was derived from: https://github.com/ziglang/zig/pull/11987

To summarize, the zig version only includes the number of commits from the last release, so, it's a bit of guesswork to figure out if a zig version would be compatible.  If you get a false positive on a "min version check" then at worst the version check passes erroneously and the project just won't compile because you need to update.  If you get a false positive on a max version check then it means your compiler is compatible but you can't compile the project because of the version check, you'll need some way to disable the version check in that case.

There are a number of ways to mitigate this and or improve the situation, we could improve our version numbers for example, maybe we make a distinction between "released versions" and "locally-built non-released versions".

**Comment by nektro** [comment]
pre-1.0 having a max version isnt worth the effort to calculate, post-1.0 having a max version would be a no-op since zig will always be forward compatible (sans maybe a policy about security fixes)

**Comment by deflock** [comment]
There should be not only minimum version but also supported versions as array or csv with AND logic:

```zig
.supported_zig_versions = .{
  ">=0.8", 
  "!0.10", // async/await is broken
  "<1.0",
}
.supported_zig_versions = ">=0.8, !0.10, <1.0"
.supported_zig_versions = "0.8" // just minimum
```

It could replace minimum field, but may complicate parsing/detecting minimum version for IDEs

**Comment by DraagrenKirneh** [comment]
Added a PR  #15268 based on the proposal from @andrewrk above.  
As per the discussion about a more complicated scheme of max versions and unsupported versions I think it is better to await and see if a more complicated system is actually needed wrt the usage of the field as it stands; ie will the compiler/build system take usage of it or will it only be a hint for editors and package loading.

**Comment by mataha** [comment]
I'm writing a project in C using Zig as a C compiler. Because it's so easy to use as one can just download it and run it, I went a step further and wrote a wrapper to automate the process so that after cloning the repository, a developer can run `zigw build` to build the project without doing anything else. Another upside of this approach is that everyone can work using the same version of the compiler, minimizing potential friction (the kind of approach that tools like [Maven](https://maven.apache.org/wrapper/) and [Gradle](https://docs.gradle.org/current/userguide/gradle_wrapper.html) advocate).

That said, I need to pull the version from somewhere - `build.zig.zon` seems like the obvious choice. My use-case prefers having one version within the string, as parsing a more complicated scheme would get tricky in shell scripts + I'm fetching a single release anyways. I even think `minimum_zig_version` is too wordy - just `zig_version` would suffice.

Playground [here](https://github.com/mataha/jigoko) (Windows only for now, i.e. Batch).

**Comment by niklaswimmer** [comment]
Do we have this issue still open because we want to also implement an error if the used Zig version is older than the specified one? Otherwise I think this is done and can be closed. We still have #10204 open too, for tracking progress/opinions on a `zig_version_max` field and any usage of those fields by the Zig compiler itself (although that issue seems a bit outdated).

**Comment by andrewrk** [comment]
Adding an error, warning, or other behavior for this field will be a follow-up proposal.
